### PR TITLE
Buffers configuration update on port speed change

### PR DESCRIPTION
**Pull request by andriymoroz-mlnx**
**REVIEW ONLY. 
Please do not commit yet.
submodules pointers needs to be updated after related PRs merged**

related PRs:
https://github.com/Azure/sonic-swss/pull/417
https://github.com/Azure/sonic-swss-common/pull/169 (merged)
https://github.com/Azure/sonic-utilities/pull/174 (merged)

**- What I did**
Config files for PG Buffers update on port speed change

**- How I did it**
See: https://github.com/Azure/SONiC/wiki/Run-Time-Buffers-Configuration-Update-design

**- How to verify it**
Buffers configuration should exist in Config DB after boot
`redis-cli -n 4 keys \* | grep BUFFER`

### CODE REVIEW

**Comment by lguohan** [comment]
@yxieca  to check?

File: device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/buffers.json.j2 [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+                {%- set roles2 = neighbor_role + '_' + switch_role -%}
+                {%- if roles1 in ports2cable -%}
+                    {%- if cable_len.append(ports2cable[roles1]) -%}{%- endif -%}
+                {%- elif roles2 in ports2cable -%}
+                    {%- if cable_len.append(ports2cable[roles2]) -%}{%- endif -%}
+                {%- endif -%}
+            {%- endif -%}
+        {%- endif -%}
+    {%- endfor -%}
+    {%- if cable_len -%}
```
Comment (yxieca) [inline device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/buffers.json.j2]: I think you need to test if the list is non-empty, but this test seems to be testing if the list has been defined. Which is defined in line 17 above.

File: device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/buffers.json.j2 [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        {{ cable_len.0 }}
```
Comment (yxieca) [inline device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/buffers.json.j2]: Under what circumstances would there be more than one length in the list?

File: device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/buffers.json.j2 [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+{%- macro cable_length(port_name) -%}
+    {%- set cable_len = [] -%}
+    {%- for local_port in DEVICE_NEIGHBOR -%}
+        {%- if local_port == port_name -%}
+            {%- if DEVICE_NEIGHBOR_METADATA[DEVICE_NEIGHBOR[local_port].name] -%}
+                {%- set neighbor = DEVICE_NEIGHBOR_METADATA[DEVICE_NEIGHBOR[local_port].name] -%}
+                {%- set neighbor_role = neighbor.type -%}
+                {%- set roles1 = switch_role + '_' + neighbor_role %}
+                {%- set roles2 = neighbor_role + '_' + switch_role -%}
+                {%- if roles1 in ports2cable -%}
```
Comment (yxieca) [inline device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/buffers.json.j2]: Do we have enforcement elsewhere to make sure that the switch_role and neighbor_role's case is as expected?

If we cannot guarantee the case, then we might want to cast the string to upper or lower case, just in case?

Same question for line 27 below.

File: device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/pg_profile_lookup.ini [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  10000  40m  36864  18432 18432  1
+  25000  40m  39936  18432 21504  1
+  40000  40m  41984  18432 23552  1
+  50000  40m  41984  18432 23552  1
+ 100000  40m  54272  18432 35840  1
+  10000  300m 49152  18432 30720  1
+  25000  300m 71680  18432 53248  1
+  40000  300m 94208  18432 75776  1
+  50000  300m 94208  18432 75776  1
+ 100000  300m 184320 18432 165888 1
```
Comment (yxieca) [inline device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/pg_profile_lookup.ini]: I didn't see matching code using this look up table. Who is using it where and how?

File: device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/buffers.json.j2 (hunk 1 above)
Comment (andriymoroz-mlnx) [reply device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/buffers.json.j2]: right, thanks. I already implemented this in https://github.com/Azure/sonic-buildimage/pull/1263 need to back-merge to this PR as well

File: device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/buffers.json.j2 (hunk 2 above)
Comment (andriymoroz-mlnx) [reply device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/buffers.json.j2]: there should be no more than one value in the list. `cable_len.0` is the way to get first (and only) element of the list. All that weird usage of the cable_len list here is the workaround for jinja templates limitation for variables visibility in blocks. Only list or dictionary can be  modified in block and maintain that value outside the block.

File: device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/pg_profile_lookup.ini (hunk 4 above)
Comment (andriymoroz-mlnx) [reply device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/pg_profile_lookup.ini]: it is passed as an argument for `buffermgrd`: https://github.com/Azure/sonic-buildimage/pull/1250/files#diff-849a3366595afffe34a13f4a543dca56

File: device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/buffers.json.j2 (hunk 3 above)
Comment (andriymoroz-mlnx) [reply device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/buffers.json.j2]: no, we don't have case enforcement. Agree, need to convert to common case before the comparison

**Comment by liatgrozovik** [comment]
following a session on that PR, it is agreed that the tool is nice to have and is not blocking the PR to be merged. But a wiki should be provided inorder to indicate how to convert an old configuration file to the new format. This should be done on another PR.

@andriymoroz-mlnx let's fix the build issues and move on to merge.

**Comment by lguohan** [comment]
@liatgrozovik , pending on Azure/sonic-swss#417, other PRs are merged.
